File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/CBuilder/EditorsInPlaceDemo/EditorsInPlaceDemoCarInfo.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "EditorsInPlaceDemoCarInfo.h"
#include "EditorsInPlaceDemoData.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxControls"
#pragma link "cxDBVGrid"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxStyles"
#pragma link "cxVGrid"
#pragma resource "*.dfm"
TEditorsInPlaceDemoCarInfoForm *EditorsInPlaceDemoCarInfoForm;
//---------------------------------------------------------------------------
__fastcall TEditorsInPlaceDemoCarInfoForm::TEditorsInPlaceDemoCarInfoForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::GetDisplayText(
  TcxCustomEditorRowProperties *Sender, int ARecord, String &AText)
{
  AText = Sender->Caption;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::vgCarInfoStylesGetContentStyle(
  TObject *Sender, TcxCustomEditorRowProperties *AEditProp, bool AFocused,
  int ARecordIndex, TcxStyle *&AStyle)
{
  if (dynamic_cast<TcxCustomMultiEditorRow*>(AEditProp->Row) != 0)
    if (((TcxMultiEditorRow*)AEditProp->Row)->Properties->Editors->Items[0] == AEditProp)
      AStyle = EditorsInPlaceDemoDataDM->styCaption;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::vgCarInfoLeftVisibleRecordIndexChanged(
  TObject *Sender)
{
  FEditValue = vgCarInfoID->Properties->Values[vgCarInfo->LeftVisibleRecord];
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::cxButtonClick(TObject *Sender)
{
  ClosePopup((bool)((TcxButton*)Sender)->Tag);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::OnEditPropertiesButtonClick(
  TObject *Sender)
{
  ShowMessage("Button click event handler");
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::ClosePopup(bool AAccepted)
{
  if (FPopupEdit == NULL) return;
  FAccepted = AAccepted;
  FPopupEdit->DroppedDown = false;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceDemoCarInfoForm::InitPopupPanel(Variant ACarID)
{
  vgCarInfo->DataController->KeyFieldNames = "ID";
  vgCarInfo->DataController->LocateByKey(ACarID);
  FAccepted = false;
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/Delphi/EditorsInPlaceDemo/EditorsInPlaceDemoCarInfo.pas
unit EditorsInPlaceDemoCarInfo;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, EditorsInPlaceDemoData, cxStyles, cxMaskEdit, cxMemo,
  cxBlobEdit, cxCurrencyEdit, cxInplaceContainer, cxControls,
  ExtCtrls, cxRadioGroup, cxImage, cxCheckBox, cxImageComboBox,
  cxDropDownEdit, cxTextEdit, cxHyperLinkEdit, cxMRUEdit, ImgList,
  cxGraphics, cxCustomData, cxLookAndFeelPainters, StdCtrls, cxButtons,
  cxDBVGrid, cxFilter, cxData, cxEdit, DB, cxDBData, cxClasses, cxVGrid, CarsData;

type
  TEditorsInPlaceDemoCarInfoForm = class(TForm)
    pnlCarInfo: TPanel;
    vgCarInfo: TcxDBVerticalGrid;
    Panel2: TPanel;
    btnOk: TcxButton;
    btnCancel: TcxButton;
    vgCarInfoCar: TcxDBMultiEditorRow;
    vgCarInfoCategory: TcxDBMultiEditorRow;
    vgCarInfoImage: TcxDBEditorRow;
    vgCarInfoLargeImageAndPrice: TcxDBMultiEditorRow;
    vgCarInfoEngine: TcxDBMultiEditorRow;
    vgCarInfoTransmission: TcxDBMultiEditorRow;
    vgCarInfoMPG: TcxDBMultiEditorRow;
    vgCarInfoWebSite: TcxDBMultiEditorRow;
    vgCarInfoDescription: TcxDBEditorRow;
    vgCarInfoID: TcxDBEditorRow;
    procedure GetDisplayText(
      Sender: TcxCustomEditorRowProperties; ARecord: Integer;
      var AText: String);
    procedure vgCarInfoStylesGetContentStyle(Sender: TObject;
      AEditProp: TcxCustomEditorRowProperties; AFocused: Boolean;
      ARecordIndex: Integer; var AStyle: TcxStyle);
    procedure vgCarInfoLeftVisibleRecordIndexChanged(Sender: TObject);
    procedure cxButtonClick(Sender: TObject);
    procedure OnEditPropertiesButtonClick(
      Sender: TObject);
  private
    FPopupEdit: TcxPopupEdit;
    FAccepted: Boolean;
    FEditValue: Variant;
    procedure ClosePopup(AAccepted: Boolean);
  public
    procedure InitPopupPanel(ACarID: Variant);
    property PopupEdit: TcxPopupEdit read FPopupEdit write FPopupEdit;
    property Accepted: Boolean read FAccepted write FAccepted;
    property EditValue: Variant read FEditValue;
  end;

var
  EditorsInPlaceDemoCarInfoForm: TEditorsInPlaceDemoCarInfoForm;

implementation

{$R *.dfm}

procedure TEditorsInPlaceDemoCarInfoForm.InitPopupPanel(ACarID: Variant);
begin
  vgCarInfo.DataController.KeyFieldNames := 'ID';
  vgCarInfo.DataController.LocateByKey(ACarID);
  FAccepted := False;
end;

procedure TEditorsInPlaceDemoCarInfoForm.ClosePopup(AAccepted: Boolean);
begin
  if FPopupEdit = nil then Exit;
  FAccepted := AAccepted;
  FPopupEdit.DroppedDown := False;
end;

procedure TEditorsInPlaceDemoCarInfoForm.GetDisplayText(
  Sender: TcxCustomEditorRowProperties; ARecord: Integer;
  var AText: String);
begin
  AText := Sender.Caption;
end;

procedure TEditorsInPlaceDemoCarInfoForm.vgCarInfoStylesGetContentStyle(
  Sender: TObject; AEditProp: TcxCustomEditorRowProperties;
  AFocused: Boolean; ARecordIndex: Integer; var AStyle: TcxStyle);
begin
  if AEditProp.Row is TcxCustomMultiEditorRow then
    if TcxMultiEditorRow(AEditProp.Row).Properties.Editors[0] = AEditProp then
      AStyle := EditorsInPlaceDemoDataDM.styCaption;
end;

procedure TEditorsInPlaceDemoCarInfoForm.vgCarInfoLeftVisibleRecordIndexChanged(
  Sender: TObject);
begin
  FEditValue := vgCarInfoID.Properties.Values[vgCarInfo.LeftVisibleRecord];
end;

procedure TEditorsInPlaceDemoCarInfoForm.cxButtonClick(Sender: TObject);
begin
  ClosePopup(Boolean(TcxButton(Sender).Tag));
end;

procedure TEditorsInPlaceDemoCarInfoForm.OnEditPropertiesButtonClick(
  Sender: TObject);
begin
  ShowMessage('Button click event handler');
end;

end.

